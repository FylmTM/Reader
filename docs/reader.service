[Unit]
Description=Reader
After=network.target
StartLimitIntervalSec=0

[Service]
Type=simple
Restart=always
RestartSec=1
User=root
Environment=ROCKET_PORT=80
Environment=RUST_LOG=warn
WorkingDirectory=/root/reader
ExecStart=/root/reader/reader

[Install]
WantedBy=multi-user.target
