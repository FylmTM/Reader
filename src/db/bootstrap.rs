use rusqlite::{Connection, NO_PARAMS};

const SCHEMA_VERSION: u16 = 1;
// language=SQLite
const SCHEMA_VERSION_QUERY: &str = "
create table if not exists schema_version
(
    id      integer primary key,
    version integer
);

insert or ignore into schema_version(id, version)
values (1, 0);
";

pub fn initialize_schema(conn: &Connection) {
    conn.execute_batch(SCHEMA_VERSION_QUERY)
        .expect("Failed to create schema_version table.");

    let mut current_version: u16 = conn
        .query_row(
            // language=SQLite
            "select version from schema_version",
            NO_PARAMS,
            |row| row.get(0),
        )
        .expect("Failed to retrieve current schema version.");

    let version_before = current_version;

    info!(
        "Current schema version=V{}, required=V{}.",
        current_version, SCHEMA_VERSION
    );

    if current_version == 0 {
        info!("Applying V1 migration.");
        conn.execute_batch(include_str!("migrations/V1_initial.sql"))
            .expect("Schema creation failed.");
        current_version = 1;
    }

    if version_before != current_version {
        info!("Set current schema version to V{}.", current_version);
        conn.execute(
            // language=SQLite
            "update schema_version set version = ? where id = 1",
            &[current_version],
        )
        .expect("Failed to update current schema version.");
    }
}

pub fn initialize_fixture(conn: &Connection) {
    // language=SQLite
    let query = "
        insert or ignore into users(id, username, api_key)
        values (1, 'me', 'api_key');
    ";
    conn.execute_batch(query).expect("Failed to apply fixture.");
}
