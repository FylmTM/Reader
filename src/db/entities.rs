#[derive(Serialize, Deserialize)]
pub struct Post {
    pub id: i64,
    pub title: String,
    pub body: String,
}
