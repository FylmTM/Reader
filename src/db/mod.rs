use rusqlite::{Connection, NO_PARAMS};

pub use self::entities::Post;
use crate::error::Error;

mod entities;

pub fn initialize(conn: &Connection) {
    conn.execute(include_str!("queries/schema.sql"), NO_PARAMS)
        .expect("schema created");
}

pub fn find_posts(conn: &Connection) -> Result<Vec<Post>, Error> {
    let mut statement = conn.prepare("SELECT id, title, body FROM posts")?;
    let result = statement.query_map(NO_PARAMS, |row| {
        Ok(Post {
            id: row.get(0)?,
            title: row.get(1)?,
            body: row.get(2)?,
        })
    })?;

    let mut posts = Vec::new();
    for post in result {
        posts.push(post?);
    }
    Ok(posts)
}
