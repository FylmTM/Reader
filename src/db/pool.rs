use std::ops::{Deref, DerefMut};

use rocket::http::Status;
use rocket::request::FromRequest;
use rocket::{Outcome, Request, State};

use crate::db;
use crate::error::Error as E;

pub type Pool = r2d2::Pool<r2d2_sqlite::SqliteConnectionManager>;

pub struct PoolConnection {
    pub connection: r2d2::PooledConnection<r2d2_sqlite::SqliteConnectionManager>,
}

impl Deref for PoolConnection {
    type Target = db::Connection;

    fn deref(&self) -> &Self::Target {
        &self.connection
    }
}

impl DerefMut for PoolConnection {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.connection
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for PoolConnection {
    type Error = E;

    fn from_request(request: &'a Request<'r>) -> rocket::request::Outcome<Self, E> {
        let pool = request
            .guard::<State<db::Pool>>()
            .expect("Failed to get pool.");

        match pool.clone().get() {
            Ok(connection) => Outcome::Success(PoolConnection { connection }),
            Err(err) => {
                error!("Failed to acquire connection.");
                Outcome::Failure((Status::InternalServerError, E::Pool(err)))
            }
        }
    }
}
