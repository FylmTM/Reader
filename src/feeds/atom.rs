use crate::db::{Feed, MediaType, Post};
use crate::error::{ApplicationError, Error, Result};
use crate::feeds::utils::{clean_to_safe_html, clean_to_text, find_image, get_url_content};

pub fn update(feed: &Feed) -> Result<Vec<Post>> {
    let content = get_url_content(&feed.feed)?;
    let posts = parse(&content)?;
    Ok(posts)
}

fn parse(content: &str) -> Result<Vec<Post>> {
    let feed = content.parse::<atom_syndication::Feed>()?;
    let mut posts = Vec::with_capacity(feed.entries().len());

    for entry in feed.entries() {
        let unique_id = entry.id().to_string();
        let link = entry
            .links()
            .iter()
            .find(|link| link.rel() == "alternate")
            .or_else(|| entry.links().get(0))
            .map(|link| link.href())
            .ok_or(Error::App(ApplicationError::AtomParsingMissingEntryLink))?
            .to_string();
        let title = entry.title().to_string();
        let date = parse_date(entry.published().unwrap_or(entry.updated()))?;
        let summary = entry.summary().map(clean_to_text);
        let content = entry
            .content()
            .and_then(|content| content.value())
            .or(entry.summary())
            .map(|content| clean_to_safe_html(content, &link));
        let comments_link = None;

        let (media_type, media_link) = {
            content
                .as_ref()
                .and_then(|content| find_image(content))
                .map(|image| {
                    (
                        Some(MediaType {
                            mime: "image/*".to_string(),
                        }),
                        Some(image),
                    )
                })
                .unwrap_or((None, None))
        };

        let post = Post {
            id: -1, // Post is not retrieved from database
            unique_id,
            link,
            title,
            date,
            summary,
            content,
            media_type,
            media_link,
            comments_link,
        };
        posts.push(post);
    }

    // Items are sorted from newest to oldest, but for database insertion
    // we need them sorted from oldest to newest.
    posts.reverse();

    Ok(posts)
}

fn parse_date(date: &str) -> Result<chrono::DateTime<chrono::Utc>> {
    let result =
        chrono::DateTime::parse_from_rfc3339(date).map(|date| chrono::DateTime::from(date))?;

    Ok(result)
}

#[cfg(test)]
mod tests {
    use crate::feeds::atom::parse;

    #[test]
    fn test_parse_atom_github_rocket_releases() {
        let content = include_str!("../../static/mock/atom/github_rocket_releases.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_golang() {
        let content = include_str!("../../static/mock/atom/golang.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_julia_evans() {
        let content = include_str!("../../static/mock/atom/julia_evans.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_rust() {
        let content = include_str!("../../static/mock/atom/rust.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }
}
