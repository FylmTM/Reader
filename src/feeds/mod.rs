use std::thread;
use std::time::Duration;

use crate::db::{self, Queries};
use crate::error::{ApplicationError, Error, Result};

pub mod atom;
pub mod rss;
pub mod utils;

pub fn initialize(pool: &db::Pool, interval: u64) {
    let pool = pool.clone();

    info!("Schedule feeds update every {} minutes.", interval);
    thread::spawn(move || loop {
        thread::sleep(Duration::from_secs(5));
        loop {
            let mut conn = pool.get().expect("Failed to acquire db connection");
            update(&mut conn);

            info!("Next feed update in {} minutes.", interval);
            thread::sleep(Duration::from_secs(60 * interval));
        }
    });
}

pub fn update(mut conn: &mut db::Connection) {
    info!("Feeds update started.");
    match conn.find_feeds() {
        Ok(feeds) => {
            for feed in feeds {
                match update_feed(&mut conn, &feed) {
                    Ok(_) => debug!("{:?} feed updated.", feed),
                    Err(error) => error!("{:?} feed failed to update: {:?}", feed, error),
                };
            }
        }
        Err(error) => {
            error!("Failed to retrieve feeds: {:?}.", error);
        }
    }
    info!("Feeds update completed.");
}

pub fn update_feed(conn: &mut db::Connection, feed: &db::Feed) -> Result<()> {
    debug!("{:?} update started.", feed);

    let user_ids = conn.find_user_ids_by_feed(feed.id)?;
    if user_ids.is_empty() {
        debug!("{:?} does not belong to any user, skip update.", feed);
        return Ok(());
    }

    let posts = match feed.kind {
        db::FeedKind::RSS => rss::update(&feed),
        db::FeedKind::Atom => atom::update(&feed),
    }?;

    save_posts(conn, feed, &posts, &user_ids)?;

    Ok(())
}

fn save_posts(
    conn: &mut db::Connection,
    feed: &db::Feed,
    posts: &Vec<db::Post>,
    user_ids: &Vec<db::UserId>,
) -> Result<()> {
    for post in posts {
        let tx = conn.transaction()?;
        match tx.save_post(feed.id, &post) {
            Ok(post_id) => {
                debug!("{:?} saved.", post);
                for user_id in user_ids {
                    tx.save_user_post(*user_id, post_id)?;
                }
                tx.commit()?;
            }
            Err(Error::App(ApplicationError::QueryEntityAlreadyExists)) => {
                debug!("{:?} already exists.", &post)
            }
            Err(error) => return Err(error),
        }
    }
    Ok(())
}
