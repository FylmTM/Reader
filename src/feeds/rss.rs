use crate::db::{self, Queries};
use crate::error::{ApplicationError, Error, Result};
use crate::feeds::utils;

pub fn update(conn: &mut db::Connection, feed: &db::Feed) -> Result<()> {
    let user_ids = conn.find_user_ids_by_feed(feed.id)?;
    if user_ids.is_empty() {
        debug!("{:?} does not belong to any user, skip update.", feed);
        return Ok(());
    }
    let content = get(&feed.feed)?;
    let posts = parse(&content)?;

    for post in posts {
        let tx = conn.transaction()?;

        match tx.save_post(feed.id, &post) {
            Ok(post_id) => {
                debug!("{:?} saved.", post);
                for user_id in &user_ids {
                    tx.save_user_post(*user_id, post_id)?;
                }
                tx.commit()?;
            }
            Err(Error::App(ApplicationError::QueryEntityAlreadyExists)) => {
                debug!("{:?} already exists.", &post)
            }
            Err(error) => return Err(error),
        }
    }

    Ok(())
}

fn get(link: &String) -> Result<String> {
    let mut response = reqwest::get(link)?;

    if response.status() != reqwest::StatusCode::OK {
        error!(
            "Failed to retrieve feed={} with status={}",
            link,
            response.status()
        );
        return Err(Error::App(ApplicationError::RSSFailedToRetrieve(
            link.clone(),
        )));
    }

    return Ok(response.text()?);
}

fn parse(content: &String) -> Result<Vec<db::Post>> {
    let channel = rss::Channel::read_from(content.as_bytes())?;

    let mut posts = Vec::with_capacity(channel.items().len());

    for item in channel.items() {
        let link = item
            .link()
            .ok_or(Error::App(ApplicationError::RSSParsingMissingItemLink))?
            .to_string();
        let title = item
            .title()
            .ok_or(Error::App(ApplicationError::RSSParsingMissingItemLink))?
            .to_string();
        let date = match item.pub_date() {
            Some(date) => parse_date(date)?,
            None => chrono::DateTime::from(chrono::Utc::now()),
        };

        let summary = item.description().map(utils::clean_to_text);
        let content = item
            .content()
            .or(item.description())
            .map(|content| utils::clean_to_safe_html(content, &link));
        let comments_link = item.comments().map(String::from);

        let media_type = item.enclosure().and_then(|enclosure| {
            let mime_type = enclosure.mime_type();
            if mime_type.is_empty() {
                return None;
            }
            if mime_type.parse::<mime::Mime>().is_err() {
                return None;
            }
            Some(db::MediaType {
                mime: mime_type.to_string(),
            })
        });
        let media_link = item.enclosure().and_then(|enclosure| {
            let url = enclosure.url();
            if url.is_empty() {
                None
            } else {
                Some(url.to_string())
            }
        });

        let post = db::Post {
            id: -1, // Post is not retrieved from database
            link,
            title,
            date,
            summary,
            content,
            media_type,
            media_link,
            comments_link,
        };
        posts.push(post);
    }

    // Items are sorted from newest to oldest, but for database insertion
    // we need them sorted from oldest to newest.
    posts.reverse();
    Ok(posts)
}

fn parse_date(date: &str) -> Result<chrono::DateTime<chrono::Utc>> {
    let result = chrono::DateTime::parse_from_rfc2822(date)
        .or_else(|_| {
            // chrono consider "-0000" timezone to be "missing"
            // We will blindly replace it with +0000 if it exists, and try to parse again.
            let date = date.replace("-0000", "+0000");
            chrono::DateTime::parse_from_rfc2822(&date)
        })
        .map(|date| chrono::DateTime::from(date))?;
    Ok(result)
}

#[cfg(test)]
mod tests {
    use crate::feeds::rss::{parse, parse_date};

    #[test]
    fn test_parse_rss_hacker_news() {
        let content = include_str!("../../static/mock/rss/hacker_news.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_kotlin() {
        let content = include_str!("../../static/mock/rss/kotlin.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_react() {
        let content = include_str!("../../static/mock/rss/react.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_stopgame() {
        let content = include_str!("../../static/mock/rss/stopgame.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_xkcd() {
        let content = include_str!("../../static/mock/rss/xkcd.xml").to_string();
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_date_handle_negative_zero_timezone() {
        let expected = chrono::DateTime::parse_from_rfc3339("2019-12-04T05:00:00+00:00").unwrap();
        assert_eq!(
            expected,
            parse_date("Wed, 04 Dec 2019 05:00:00 -0000").unwrap()
        );
    }
}
