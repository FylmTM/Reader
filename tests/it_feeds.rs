#![feature(proc_macro_hygiene, decl_macro)]
#[macro_use]
extern crate rocket;
#[macro_use]
extern crate insta;

use reader::db::Queries;
use reader::feeds;

pub mod common;

const USER_ID: i64 = 1;
const RSS_HACKER_NEWS_ID: i64 = 1;

#[test]
fn test_rss_update() {
    common::mock_server::start();

    let mut conn = common::db();
    let feed = conn.find_feed(RSS_HACKER_NEWS_ID).unwrap();

    // Empty
    assert_json_snapshot!(
        "rss_update_before",
        conn.find_posts(USER_ID, None, None, None, None, Some(RSS_HACKER_NEWS_ID))
            .unwrap()
    );

    // Updated
    feeds::update_feed(&mut conn, &feed).unwrap();
    assert_json_snapshot!(
        "rss_update_after",
        conn.find_posts(USER_ID, None, None, None, None, Some(RSS_HACKER_NEWS_ID))
            .unwrap()
    );

    // No changes
    feeds::update_feed(&mut conn, &feed).unwrap();
    assert_json_snapshot!(
        "rss_update_after",
        conn.find_posts(USER_ID, None, None, None, None, Some(RSS_HACKER_NEWS_ID))
            .unwrap()
    );
}
