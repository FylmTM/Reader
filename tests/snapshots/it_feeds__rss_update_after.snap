---
source: tests/it_feeds.rs
expression: "conn.find_posts(USER_ID, None, None, None, None,\n                Some(RSS_HACKER_NEWS_ID)).unwrap()"
---
{
  "has_next_page": false,
  "items": [
    {
      "id": 30,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://blog.cynthia.re/post/nintendo-switch-ethernet-switch",
      "title": "The Nintendo Switch Switch",
      "date": "2019-12-09T19:46:52Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21745731"
    },
    {
      "id": 29,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://en.wikipedia.org/wiki/Transputer",
      "title": "Transputer",
      "date": "2019-12-09T21:22:57Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21746739"
    },
    {
      "id": 28,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://randomascii.wordpress.com/2019/12/08/on2-again-now-in-wmi/",
      "title": "O(n^2), again, now in Windows Management Instrumentation",
      "date": "2019-12-09T15:30:35Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21743424"
    },
    {
      "id": 27,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.wsj.com/articles/american-factories-demand-white-collar-education-for-blue-collar-work-11575907185",
      "title": "College-educated workers are taking over the American factory floor",
      "date": "2019-12-09T19:43:30Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21745704"
    },
    {
      "id": 26,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.nytimes.com/2019/12/08/obituaries/caroll-spinney-dead.html",
      "title": "Caroll Spinney has died",
      "date": "2019-12-08T19:31:16Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21737697"
    },
    {
      "id": 25,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://signal.org/blog/signal-private-group-system/",
      "title": "Technology Preview: Signal Private Group System",
      "date": "2019-12-09T17:10:11Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21744274"
    },
    {
      "id": 24,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.sametab.com/blog/operations-and-internal-communication-strategies-for-effective-ceos",
      "title": "Operations for Effective CEOs",
      "date": "2019-12-09T14:11:22Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21742739"
    },
    {
      "id": 23,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.wsj.com/articles/if-bitcoin-looks-like-it-isnt-trading-its-because-it-isnt-11575637204",
      "title": "If Bitcoin Looks Like It Isn’t Trading, It’s Because It Isn’t",
      "date": "2019-12-09T21:06:13Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21746582"
    },
    {
      "id": 22,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.quantamagazine.org/how-simple-rules-bootstrap-the-laws-of-physics-20191209/",
      "title": "Simple Rules ‘Bootstrap’ the Laws of Physics",
      "date": "2019-12-09T16:20:49Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21743847"
    },
    {
      "id": 21,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://web.archive.org/web/20030812081713/http://klausler.com/cnotes.txt",
      "title": "Some things every C programmer should know about C (2002)",
      "date": "2019-12-09T15:59:55Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21743696"
    },
    {
      "id": 20,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://emacsair.me/2017/09/01/magit-walk-through/",
      "title": "A walk through the Magit interface",
      "date": "2019-12-07T10:55:30Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21729597"
    },
    {
      "id": 19,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "http://scholar.harvard.edu/files/mickens/files/towashitallaway.pdf",
      "title": "To Wash It All Away (2014) [pdf]",
      "date": "2019-12-09T18:00:56Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21744785"
    },
    {
      "id": 18,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://queue.acm.org/detail.cfm?id=3341158",
      "title": "Must and Must Not: On writing documentation",
      "date": "2019-12-09T14:53:02Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21743100"
    },
    {
      "id": 17,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=e7096c131e5161fa3b8e52a650d7719d2857adfd",
      "title": "WireGuard is in net-next",
      "date": "2019-12-09T08:21:03Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21741133"
    },
    {
      "id": 16,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.atlasobscura.com/articles/cool-bar-montana",
      "title": "For Sale: The Only Bar in a 14-Person Montana Town",
      "date": "2019-12-09T19:40:03Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21745673"
    },
    {
      "id": 15,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.theatlantic.com/ideas/archive/2019/11/why-child-care-so-expensive/602599",
      "title": "Why Child Care Is So Expensive",
      "date": "2019-12-09T18:41:48Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21745124"
    },
    {
      "id": 14,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.bloomberg.com/news/articles/2019-12-08/repo-blowup-was-fueled-by-big-banks-and-hedge-funds-bis-says",
      "title": "Repo Blowup Was Fueled by Big Banks and Hedge Funds, BIS Says",
      "date": "2019-12-09T14:05:49Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21742702"
    },
    {
      "id": 13,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://github.com/infostreams/db",
      "title": "Show HN: Version Control for Databases",
      "date": "2019-12-09T13:06:35Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21742222"
    },
    {
      "id": 12,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.wsj.com/articles/five-cities-account-for-vast-majority-of-growth-in-tech-jobs-study-finds-11575867660",
      "title": "Five cities account for vast majority of growth in U.S. tech jobs: study",
      "date": "2019-12-09T14:56:35Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21743129"
    },
    {
      "id": 11,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://github.com/ecaron/smart-nightlight-manager",
      "title": "Show HN: My in-network Node app for controlling my kids' nightlights",
      "date": "2019-12-08T22:52:12Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21739000"
    },
    {
      "id": 10,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://jamesmunns.com/blog/fmt-unreasonably-expensive/",
      "title": "Formatting Is Unreasonably Expensive for Embedded Rust",
      "date": "2019-12-09T15:41:11Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21743513"
    },
    {
      "id": 9,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://fivebooks.com/best-books/science-2019-barbara-kiser/",
      "title": "Science Books of 2019",
      "date": "2019-12-09T17:59:35Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21744772"
    },
    {
      "id": 8,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "http://web.archive.org/web/20040929041451/http://www.shrovetuesdayobserved.com/flight.html",
      "title": "If All Stories Were Written Like Science Fiction Stories (2004)",
      "date": "2019-12-09T21:05:24Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21746574"
    },
    {
      "id": 7,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.thestar.com.my/news/regional/2019/12/09/breaking-news-russia-banned-from-2020-olympics",
      "title": "Russia Banned from 2020 Olympics",
      "date": "2019-12-09T21:16:07Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21746679"
    },
    {
      "id": 6,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.bbc.com/news/technology-50713723",
      "title": "E-scooter company goes bust after spending big on Facebook ads",
      "date": "2019-12-09T19:13:16Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21745389"
    },
    {
      "id": 5,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://github.com/Risk3sixty-Labs/geoapi",
      "title": "Show HN: Tiny, fast, and free API to geolocate IP addresses",
      "date": "2019-12-09T12:18:47Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21741991"
    },
    {
      "id": 4,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.newyorker.com/magazine/2019/12/16/how-william-gibson-keeps-his-science-fiction-real",
      "title": "How William Gibson Keeps His Science Fiction Real",
      "date": "2019-12-09T18:35:49Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21745074"
    },
    {
      "id": 3,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.tmakiguchi.org/geographer/asgeographer/geographyhuman.html",
      "title": "Tsunebaro Makaguchi – Education Through Interest",
      "date": "2019-12-08T06:21:34Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21734605"
    },
    {
      "id": 2,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.vitavonni.de/blog/201503/2015031201-the-sad-state-of-sysadmin-in-the-age-of-containers.html",
      "title": "The sad state of sysadmin in the age of containers (2015)",
      "date": "2019-12-09T10:05:26Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21741504"
    },
    {
      "id": 1,
      "category_id": 1,
      "feed_id": 1,
      "feed_title": "Hacker News",
      "is_read": false,
      "read_date": null,
      "is_read_later": false,
      "link": "https://www.reuters.com/article/us-eu-batteries/eu-approves-3-2-billion-euro-state-aid-for-battery-research-idUSKBN1YD0WJ",
      "title": "EU approves 3.2B euro state aid for battery research",
      "date": "2019-12-09T21:15:22Z",
      "summary": "Comments",
      "media_link": null,
      "comments_link": "https://news.ycombinator.com/item?id=21746673"
    }
  ]
}
